<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>gff.c File Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.2.15 -->
<center>
<a class="qindex" href="index.html">Main Page</a> &nbsp; <a class="qindex" href="modules.html">Modules</a> &nbsp; <a class="qindex" href="annotated.html">Compound List</a> &nbsp; <a class="qindex" href="files.html">File List</a> &nbsp; <a class="qindex" href="functions.html">Compound Members</a> &nbsp; <a class="qindex" href="globals.html">File Members</a> &nbsp; </center>
<hr><h1>gff.c File Reference</h1>Reading and writing of sequence features in General Feature Format (GFF). 
<a href="#_details">More...</a>
<p>
<table border=0 cellpadding=0 cellspacing=0>
<tr><td colspan=2><br><h2>Functions</h2></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a0">gff_read_set</a> (FILE *F)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Read a set of features from a file and return a newly allocated <a class="el" href="structGFF__Set.html">GFF_Set</a> object.</em> <a href="#a0">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Feature.html">GFF_Feature</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a1">gff_new_feature</a> (String *seqname, String *source, String *feature, int start, int end, double score, char strand, int frame, String *attribute, int score_is_null)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create new <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object with specified attributes.</em> <a href="#a1">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Feature.html">GFF_Feature</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a2">gff_new_feature_genomic_pos</a> (String *position, String *source, String *feature, double score, int frame, String *attribute, int score_is_null)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create a new <a class="el" href="structGFF__Feature.html">GFF_Feature</a> from a genomic position string of the type used in the UCSC genome browser, e.g., chr10:102553847-102554897.</em> <a href="#a2">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a3">gff_new_set</a> ()</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create new <a class="el" href="structGFF__Set.html">GFF_Set</a> object.</em> <a href="#a3">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a4" doxytag="gff.c::gff_new_from_template"></a>
<a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a4">gff_new_from_template</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *gff)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create new <a class="el" href="structGFF__Set.html">GFF_Set</a> object, initializing with same version, source, etc. as a template <a class="el" href="structGFF__Set.html">GFF_Set</a> object.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a5">gff_new_set_init</a> (char *source, char *source_version)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create new <a class="el" href="structGFF__Set.html">GFF_Set</a> object, using typical defaults and other parameters as specified.</em> <a href="#a5">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a6" doxytag="gff.c::gff_free_set"></a>
void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a6">gff_free_set</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Free resources associated with <a class="el" href="structGFF__Set.html">GFF_Set</a> object (including all features and the set object itself).</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a7" doxytag="gff.c::gff_free_feature"></a>
void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a7">gff_free_feature</a> (<a class="el" href="structGFF__Feature.html">GFF_Feature</a> *feat)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Free resources associated with <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a8" doxytag="gff.c::gff_print_set"></a>
void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a8">gff_print_set</a> (FILE *F, <a class="el" href="structGFF__Set.html">GFF_Set</a> *set)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Output a <a class="el" href="structGFF__Set.html">GFF_Set</a> to the specified stream in GFF.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a9" doxytag="gff.c::gff_print_feat"></a>
void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a9">gff_print_feat</a> (FILE *F, <a class="el" href="structGFF__Feature.html">GFF_Feature</a> *feat)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Print an individual <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object as a GFF line.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a10" doxytag="gff.c::gff_new_feature_copy"></a>
<a class="el" href="structGFF__Feature.html">GFF_Feature</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a10">gff_new_feature_copy</a> (<a class="el" href="structGFF__Feature.html">GFF_Feature</a> *orig)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create an exact copy of a <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a11">gff_subset_range</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set, int startcol, int endcol, int reset_indices)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Create a new <a class="el" href="structGFF__Set.html">GFF_Set</a> representing the features in a particular coordinate range.</em> <a href="#a11">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a12">gff_filter_by_type</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *gff, <a class="el" href="structList.html">List</a> *include, FILE *discards_f)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Discard any feature whose feature type is *not* in the specified list.</em> <a href="#a12">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>int&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a13">gff_reverse_strand_only</a> (<a class="el" href="structList.html">List</a> *features)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Test whether a set of <a class="el" href="structGFF__Feature.html">GFF_Feature</a> objects refers to the reverse strand.</em> <a href="#a13">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a14">gff_reverse_compl</a> (<a class="el" href="structList.html">List</a> *features, int start_range, int end_range)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Adjust coordinates and strand of <a class="el" href="structGFF__Feature.html">GFF_Feature</a> objects to reflect reverse complementation of given interval of sequence.</em> <a href="#a14">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a17">gff_sort</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Sort features primarily by start position, and secondarily by end position (ascending order).</em> <a href="#a17">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a18">gff_group</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set, char *tag)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Group features by value of specified tag.</em> <a href="#a18">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a19" doxytag="gff.c::gff_ungroup"></a>
void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a19">gff_ungroup</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Remove grouping of features.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a20">gff_exon_group</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *set, char *tag)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Group contiguous features, e.g., an exon and adjacent splice sites.</em> <a href="#a20">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a21">gff_remove_overlaps</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *gff, FILE *discards_f)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Identify overlapping groups and remove all but the first one encountered.</em> <a href="#a21">More...</a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="gff_8c.html#a22">gff_fix_stops</a> (<a class="el" href="structGFF__Set.html">GFF_Set</a> *gff, char *cds_type, char *stop_type)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Adjust coords of terminal cds exons such that stop codons are included.</em> <a href="#a22">More...</a><em></em></font><br><br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Reading and writing of sequence features in General Feature Format (GFF).
<p>
 Obeys file specification at <a href="http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml.">http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml.</a>
<p>
<hr><h2>Function Documentation</h2>
<a name="a20" doxytag="gff.c::gff_exon_group"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_exon_group </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>set</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>tag</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Group contiguous features, e.g., an exon and adjacent splice sites.
<p>
If features have already been grouped (e.g., by transcript id), then subgroups are created by adding new tags. New tag values will be composed of the tag value for the "outer" group (e.g., the transcript_id) and a unique suffix indicating the "inner" group. In all cases, feature are sorted as a side effect, in a way that reflects the initial grouping, not the grouping into exons. <dl compact><dt><b>
Parameters: </b><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>set</em>&nbsp;</td><td>
Set to process </td></tr>
<tr><td valign=top><em>tag</em>&nbsp;</td><td>
Tag to use for new groups (e.g., "exon_id") </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a12" doxytag="gff.c::gff_filter_by_type"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_filter_by_type </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>gff</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap><a class="el" href="structList.html">List</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>include</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>FILE *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>discards_f</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Discard any feature whose feature type is *not* in the specified list.
<p>
<dl compact><dt><b>
Parameters: </b><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>gff</em>&nbsp;</td><td>
<a class="el" href="structGFF__Set.html">GFF_Set</a> to process </td></tr>
<tr><td valign=top><em>include</em>&nbsp;</td><td>
Feature types to include (<a class="el" href="structList.html">List</a> of Strings) </td></tr>
<tr><td valign=top><em>discards_f</em>&nbsp;</td><td>
Discarded features will be written here (if non-NULL) </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a22" doxytag="gff.c::gff_fix_stops"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_fix_stops </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>gff</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>cds_type</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>stop_type</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Adjust coords of terminal cds exons such that stop codons are included.
<p>
Assumes GFF is grouped such that at most one stop codon occurs per group and it corresponds to at most one cds. <dl compact><dt><b>
Parameters: </b><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>gff</em>&nbsp;</td><td>
Set to process </td></tr>
<tr><td valign=top><em>cds_type</em>&nbsp;</td><td>
Type indicating CDS features (e.g., "CDS") </td></tr>
<tr><td valign=top><em>stop_type</em>&nbsp;</td><td>
Type indicating stop codons (e.g., "stop_codon") </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a18" doxytag="gff.c::gff_group"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_group </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>set</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>tag</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Group features by value of specified tag.
<p>
All features with undefined values will be placed in a single group.     </td>
  </tr>
</table>
<a name="a1" doxytag="gff.c::gff_new_feature"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Feature.html">GFF_Feature</a>* gff_new_feature </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>seqname</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>source</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>feature</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>start</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>end</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>double&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>score</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>strand</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>frame</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>attribute</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>score_is_null</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Create new <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object with specified attributes.
<p>
Strings are copied by reference. Returns newly allocated <a class="el" href="structGFF__Feature.html">GFF_Feature</a> object.     </td>
  </tr>
</table>
<a name="a2" doxytag="gff.c::gff_new_feature_genomic_pos"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Feature.html">GFF_Feature</a>* gff_new_feature_genomic_pos </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>position</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>source</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>feature</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>double&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>score</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>frame</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>String *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>attribute</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>score_is_null</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Create a new <a class="el" href="structGFF__Feature.html">GFF_Feature</a> from a genomic position string of the type used in the UCSC genome browser, e.g., chr10:102553847-102554897.
<p>
A trailing '+' or '-' will be interpreted as the strand; otherwise the null strand is used. NULL is returned if the string can't be parsed.     </td>
  </tr>
</table>
<a name="a3" doxytag="gff.c::gff_new_set"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Set.html">GFF_Set</a>* gff_new_set </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="mdname1" valign="top" nowrap>&nbsp;          </td>
          <td class="md" valign="top">)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Create new <a class="el" href="structGFF__Set.html">GFF_Set</a> object.
<p>
All attributes will be left as empty strings.     </td>
  </tr>
</table>
<a name="a5" doxytag="gff.c::gff_new_set_init"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Set.html">GFF_Set</a>* gff_new_set_init </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>source</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>source_version</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Create new <a class="el" href="structGFF__Set.html">GFF_Set</a> object, using typical defaults and other parameters as specified.
<p>
Sets gff version to '2' and date to current date, and sets source and source version as specified.     </td>
  </tr>
</table>
<a name="a0" doxytag="gff.c::gff_read_set"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Set.html">GFF_Set</a>* gff_read_set </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">FILE *&nbsp;</td>
          <td class="mdname1" valign="top" nowrap>&nbsp; <em>F</em>          </td>
          <td class="md" valign="top">)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Read a set of features from a file and return a newly allocated <a class="el" href="structGFF__Set.html">GFF_Set</a> object.
<p>
Function reads until end-of-file is encountered or error occurs (aborts on error). Comments and blank lines are ignored and special "meta-data" comments are parsed (see <a href="http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml">http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml</a>). Only the first five columns of feature lines are considered required ('name', 'source', 'feature', 'start', and 'end'); subsequent fields are optional ('score', 'strand', 'frame', and 'attribute'). Default value for score, strand, and frame is null ('.') and for attribute is the empty string (''). Columns must be separated by tabs.     </td>
  </tr>
</table>
<a name="a21" doxytag="gff.c::gff_remove_overlaps"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_remove_overlaps </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>gff</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>FILE *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>discards_f</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Identify overlapping groups and remove all but the first one encountered.
<p>
Features must already be grouped. <dl compact><dt><b>
Parameters: </b><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>gff</em>&nbsp;</td><td>
Set to process </td></tr>
<tr><td valign=top><em>discards_f</em>&nbsp;</td><td>
If non-NULL, discarded features will be written here </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a14" doxytag="gff.c::gff_reverse_compl"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_reverse_compl </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structList.html">List</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>features</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>start_range</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>end_range</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Adjust coordinates and strand of <a class="el" href="structGFF__Feature.html">GFF_Feature</a> objects to reflect reverse complementation of given interval of sequence.
<p>
Also reverses order of appearance of features. The features, the start_range, and the end_range are all assumed to use the same coordinate frame. <dl compact><dt><b>
Parameters: </b><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>features</em>&nbsp;</td><td>
<a class="el" href="structList.html">List</a> of <a class="el" href="structGFF__Feature.html">GFF_Feature</a> objects </td></tr>
<tr><td valign=top><em>start_range</em>&nbsp;</td><td>
First coordinate of interval  (inclusive, 1-based indexing, as in features) </td></tr>
<tr><td valign=top><em>end_range</em>&nbsp;</td><td>
Last coordinate of interval (inclusive, 1-based indexing, as in features) </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a13" doxytag="gff.c::gff_reverse_strand_only"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> int gff_reverse_strand_only </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structList.html">List</a> *&nbsp;</td>
          <td class="mdname1" valign="top" nowrap>&nbsp; <em>features</em>          </td>
          <td class="md" valign="top">)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Test whether a set of <a class="el" href="structGFF__Feature.html">GFF_Feature</a> objects refers to the reverse strand.
<p>
Returns 1 if no features have strand equal to '+' and at least one has strand equal to '-'; otherwise returns 0.     </td>
  </tr>
</table>
<a name="a17" doxytag="gff.c::gff_sort"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void gff_sort </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname1" valign="top" nowrap>&nbsp; <em>set</em>          </td>
          <td class="md" valign="top">)&nbsp;</td>
          <td class="md" nowrap></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Sort features primarily by start position, and secondarily by end position (ascending order).
<p>
If features are grouped (see gff_group), then they will be sorted within groups, and groups will be sorted by start position of first feature     </td>
  </tr>
</table>
<a name="a11" doxytag="gff.c::gff_subset_range"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structGFF__Set.html">GFF_Set</a>* gff_subset_range </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>set</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>startcol</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>endcol</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>reset_indices</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Create a new <a class="el" href="structGFF__Set.html">GFF_Set</a> representing the features in a particular coordinate range.
<p>
Keeps features such that feat-&gt;start &gt;= startcol and feat-&gt;end &lt;= endcol.     </td>
  </tr>
</table>
<hr><address align="right"><small>Generated on Tue Jun 22 11:33:54 2004 for PHAST by
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border=0 
width=110 height=53></a>1.2.15 </small></address>
</body>
</html>
