<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>PHAST: msa.h File Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.3.9.1 -->
<div class="qindex"><a class="qindex" href="index.html">Main&nbsp;Page</a> | <a class="qindex" href="modules.html">Modules</a> | <a class="qindex" href="annotated.html">Class&nbsp;List</a> | <a class="qindex" href="files.html">File&nbsp;List</a> | <a class="qindex" href="functions.html">Class&nbsp;Members</a> | <a class="qindex" href="globals.html">File&nbsp;Members</a></div>
<div class="nav">
<a class="el" href="dir_000005.html">include</a></div>
<h1>msa.h File Reference</h1>Multiple sequence alignments. <a href="#_details">More...</a>
<p>
<table border="0" cellpadding="0" cellspacing="0">
<tr><td></td></tr>
<tr><td colspan="2"><br><h2>Classes</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">struct &nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="structMSA.html">MSA</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Multiple sequence alignment object.  <a href="structMSA.html#_details">More...</a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">struct &nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="structmsa__coord__map.html">msa_coord_map</a></td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Coordinate map, defined by a sequence/alignment pair.  <a href="structmsa__coord__map.html#_details">More...</a><br></td></tr>
<tr><td colspan="2"><br><h2>Defines</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a4" doxytag="msa.h::DEFAULT_ALPHABET"></a>
#define&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a4">DEFAULT_ALPHABET</a>&nbsp;&nbsp;&nbsp;"ACGT"</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Default alphabet, assumed throughout PHAST. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a5" doxytag="msa.h::GAP_CHAR"></a>
#define&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a5">GAP_CHAR</a>&nbsp;&nbsp;&nbsp;'-'</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Gap character, assumed throughout PHAST. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a6" doxytag="msa.h::DEFAULT_MDATA_CHARS"></a>
#define&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a6">DEFAULT_MDATA_CHARS</a>&nbsp;&nbsp;&nbsp;"*N"</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Default missing data character. <br></td></tr>
<tr><td colspan="2"><br><h2>Enumerations</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">enum &nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a70">msa_format_type</a> { <br>
&nbsp;&nbsp;<a class="el" href="msa_8h.html#a70a10">PHYLIP</a>, 
<a class="el" href="msa_8h.html#a70a11">MPM</a>, 
<a class="el" href="msa_8h.html#a70a12">FASTA</a>, 
<a class="el" href="msa_8h.html#a70a13">SS</a>, 
<br>
&nbsp;&nbsp;<a class="el" href="msa_8h.html#a70a14">LAV</a>, 
<a class="el" href="msa_8h.html#a70a15">MAF</a>
<br>
 }</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Format types.  <a href="#a70">More...</a><br></td></tr>
<tr><td colspan="2"><br><h2>Functions</h2></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a16">msa_new</a> (char **seqs, char **names, int nseqs, int length, char *alphabet)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Creates a new <a class="el" href="structMSA.html">MSA</a> object.  <a href="#a16"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a17">msa_new_from_file</a> (FILE *F, <a class="el" href="msa_8h.html#a70">msa_format_type</a> format, char *alphabet)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Creates a new alignment from the contents of the specified file, which is assumed to use the specified format.  <a href="#a17"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a18">msa_create_copy</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int suff_stats_only)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">create a copy of an <a class="el" href="structMSA.html">MSA</a>.  <a href="#a18"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a20">msa_print</a> (FILE *F, <a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="msa_8h.html#a70">msa_format_type</a> format, int pretty_print)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Prints <a class="el" href="structMSA.html">MSA</a> to file, using specified format.  <a href="#a20"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a22">msa_strip_gaps</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int gap_strip_mode)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">If gap_strip_mode is STRIP_ALL_GAPS or STRIP_ANY_GAPS, removes all columns with ALL or ANY gaps, respectively.  <a href="#a22"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a31" doxytag="msa.h::msa_get_seq_idx"></a>
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a31">msa_get_seq_idx</a> (<a class="el" href="structMSA.html">MSA</a> *msa, char *name)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Return sequence index of given sequence name or -1 if not found. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a40">msa_reverse_compl_feats</a> (<a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="structGFF__Set.html">GFF_Set</a> *feats, int *aux_data)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Reverse complement segments of an <a class="el" href="structMSA.html">MSA</a> corresponding to groups of features on the reverse strand.  <a href="#a40"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a57" doxytag="msa.h::msa_format_for_suffix"></a>
<a class="el" href="msa_8h.html#a70">msa_format_type</a>&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a57">msa_format_for_suffix</a> (char *fname)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Return format type indicated by filename suffix. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a58" doxytag="msa.h::msa_suffix_for_format"></a>
char *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a58">msa_suffix_for_format</a> (<a class="el" href="msa_8h.html#a70">msa_format_type</a> t)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Return appropriate filename suffix for format type. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="anchor" name="a61" doxytag="msa.h::msa_missing_col"></a>
int&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a61">msa_missing_col</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int ref, int pos)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Returns TRUE if alignment has missing data in all seqs but the reference seq at specified column; otherwise returns FALSE. <br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top"><a class="el" href="structlst__struct.html">List</a> *&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a62">msa_seq_indices</a> (<a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="structlst__struct.html">List</a> *seqnames)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Given a list of sequence names and/or 1-based indices, return a list of corresponding 0-based indices.  <a href="#a62"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a63">msa_mask_macro_indels</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int k, int refseq)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Mask out all alignment gaps of length greater than k by changing gap characters to missing data characters.  <a href="#a63"></a><br></td></tr>
<tr><td class="memItemLeft" nowrap align="right" valign="top">void&nbsp;</td><td class="memItemRight" valign="bottom"><a class="el" href="msa_8h.html#a64">msa_set_informative</a> (<a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="structlst__struct.html">List</a> *not_informative)</td></tr>

<tr><td class="mdescLeft">&nbsp;</td><td class="mdescRight">Set up array indicating which sequences are to be considered "informative", e.g., for phylogenetic analysis.  <a href="#a64"></a><br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Multiple sequence alignments. 
<p>
<hr><h2>Enumeration Type Documentation</h2>
<a class="anchor" name="a70" doxytag="msa.h::msa_format_type"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">enum <a class="el" href="msa_8h.html#a70">msa_format_type</a>          </td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Format types. 
<p>
<dl compact><dt><b>Enumeration values: </b></dt><dd>
<table border="0" cellspacing="2" cellpadding="0">
<tr><td valign="top"><em><a class="anchor" name="a70a10" doxytag="PHYLIP"></a>PHYLIP</em>&nbsp;</td><td>
PHYLIP format. </td></tr>
<tr><td valign="top"><em><a class="anchor" name="a70a11" doxytag="MPM"></a>MPM</em>&nbsp;</td><td>
Format used by MultiPipMaker and some of Webb Miller's older tools. </td></tr>
<tr><td valign="top"><em><a class="anchor" name="a70a12" doxytag="FASTA"></a>FASTA</em>&nbsp;</td><td>
Standard FASTA format. </td></tr>
<tr><td valign="top"><em><a class="anchor" name="a70a13" doxytag="SS"></a>SS</em>&nbsp;</td><td>
"Sufficient statistics" format, in which each unique alignment column (or tuple of columns) is represented only once, and a count is maintained of how many times it occurs </td></tr>
<tr><td valign="top"><em><a class="anchor" name="a70a14" doxytag="LAV"></a>LAV</em>&nbsp;</td><td>
lav format, used by BLASTZ </td></tr>
<tr><td valign="top"><em><a class="anchor" name="a70a15" doxytag="MAF"></a>MAF</em>&nbsp;</td><td>
Multiple Alignment Format (MAF) used by MULTIZ and TBA. </td></tr>
</table>
</dl>
    </td>
  </tr>
</table>
<hr><h2>Function Documentation</h2>
<a class="anchor" name="a18" doxytag="msa.h::msa_create_copy"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a>* msa_create_copy           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>suff_stats_only</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
create a copy of an <a class="el" href="structMSA.html">MSA</a>. 
<p>
If suff_stats_only == 1, then sequences aren't copied     </td>
  </tr>
</table>
<a class="anchor" name="a63" doxytag="msa.h::msa_mask_macro_indels"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">void msa_mask_macro_indels           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>k</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>refseq</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Mask out all alignment gaps of length greater than k by changing gap characters to missing data characters. 
<p>
If refseq is &gt; 0, the designated sequence will not be altered. This function is useful when modeling micro-indels. Warning: if <a class="el" href="structMSA.html">MSA</a> is stored only in terms of sufficient statistics, an explicit alignment will be created     </td>
  </tr>
</table>
<a class="anchor" name="a16" doxytag="msa.h::msa_new"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a>* msa_new           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">char **&nbsp;</td>
          <td class="mdname" nowrap> <em>seqs</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>char **&nbsp;</td>
          <td class="mdname" nowrap> <em>names</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>nseqs</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>length</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap> <em>alphabet</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Creates a new <a class="el" href="structMSA.html">MSA</a> object. 
<p>
Two-dimensional character arrays must be passed in for sequences and names (no new memory is allocated for them). The alphabet, however, will be copied into newly allocated memory. If the "alphabet" argument is null, the default alphabet will be used.     </td>
  </tr>
</table>
<a class="anchor" name="a17" doxytag="msa.h::msa_new_from_file"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a>* msa_new_from_file           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">FILE *&nbsp;</td>
          <td class="mdname" nowrap> <em>F</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="msa_8h.html#a70">msa_format_type</a>&nbsp;</td>
          <td class="mdname" nowrap> <em>format</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap> <em>alphabet</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Creates a new alignment from the contents of the specified file, which is assumed to use the specified format. 
<p>
If "alphabet" is NULL, default alphabet for DNA will be used. This routine will abort if the sequence contains a character not in the alphabet.     </td>
  </tr>
</table>
<a class="anchor" name="a20" doxytag="msa.h::msa_print"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">void msa_print           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">FILE *&nbsp;</td>
          <td class="mdname" nowrap> <em>F</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="msa_8h.html#a70">msa_format_type</a>&nbsp;</td>
          <td class="mdname" nowrap> <em>format</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>pretty_print</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Prints <a class="el" href="structMSA.html">MSA</a> to file, using specified format. 
<p>
The "pretty_print" option causes periods ('.') to be printed in place of characters that are identical to corresponding characters in the first sequence.     </td>
  </tr>
</table>
<a class="anchor" name="a40" doxytag="msa.h::msa_reverse_compl_feats"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">void msa_reverse_compl_feats           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>feats</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int *&nbsp;</td>
          <td class="mdname" nowrap> <em>aux_data</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Reverse complement segments of an <a class="el" href="structMSA.html">MSA</a> corresponding to groups of features on the reverse strand. 
<p>
Adjusts the coordinates in the <a class="el" href="structGFF__Set.html">GFF_Set</a> accordingly. This function can be used to ensure that sites in strand-specific categories (e.g., 1st codon position, intron) are oriented consistently. It can be useful in phylogenetic analysis and in training the transition probabilities of a phylo-HMM. Features are assumed already to have been grouped as desired, and groups are assumed to be nonoverlapping (see gff_group and gff_remove_overlaps). Strandedness is tested using gff_reverse_strand_only. The <a class="el" href="structGFF__Set.html">GFF_Set</a> is assumed to use the coordinate frame of the alignment. <dl compact><dt><b>Parameters: </b></dt><dd>
<table border="0" cellspacing="2" cellpadding="0">
<tr><td valign="top"><em>msa</em>&nbsp;</td><td>
Alignment object. If NULL, only the <a class="el" href="structGFF__Set.html">GFF_Set</a> (and optionally aux_data) will be altered </td></tr>
<tr><td valign="top"><em>feats</em>&nbsp;</td><td>
Set of features </td></tr>
<tr><td valign="top"><em>aux_data</em>&nbsp;</td><td>
Auxiliary array of site-specific integers (e.g., gap patterns) to be kept in sync with the alignment and/or <a class="el" href="structGFF__Set.html">GFF_Set</a> </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a class="anchor" name="a62" doxytag="msa.h::msa_seq_indices"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"><a class="el" href="structlst__struct.html">List</a>* msa_seq_indices           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="structlst__struct.html">List</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>seqnames</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Given a list of sequence names and/or 1-based indices, return a list of corresponding 0-based indices. 
<p>
Warn if a name has no match. Useful in converting command-line arguments     </td>
  </tr>
</table>
<a class="anchor" name="a64" doxytag="msa.h::msa_set_informative"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">void msa_set_informative           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap><a class="el" href="structlst__struct.html">List</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>not_informative</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Set up array indicating which sequences are to be considered "informative", e.g., for phylogenetic analysis. 
<p>
<dl compact><dt><b>Parameters: </b></dt><dd>
<table border="0" cellspacing="2" cellpadding="0">
<tr><td valign="top"><em>msa</em>&nbsp;</td><td>
Alignment </td></tr>
<tr><td valign="top"><em>not_informative</em>&nbsp;</td><td>
List of names of sequences *not* to be considered informative </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a class="anchor" name="a22" doxytag="msa.h::msa_strip_gaps"></a><p>
<table class="mdTable" width="100%" cellpadding="2" cellspacing="0">
  <tr>
    <td class="mdRow">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top">void msa_strip_gaps           </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap> <em>msa</em>, </td>
        </tr>
        <tr>
          <td class="md" nowrap align="right"></td>
          <td class="md"></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap> <em>gap_strip_mode</em></td>
        </tr>
        <tr>
          <td class="md"></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>
      </table>
    </td>
  </tr>
</table>
<table cellspacing="5" cellpadding="0" border="0">
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
If gap_strip_mode is STRIP_ALL_GAPS or STRIP_ANY_GAPS, removes all columns with ALL or ANY gaps, respectively. 
<p>
Otherwise, assumes a projection* is desired onto the sequence whose index is gap_strip_mode (indexing starts with 1). Changes are made to original alignment. Gaps are expected to be represented by GAP_CHAR. If msa-&gt;categories is non-NULL, will be adjusted accordingly.     </td>
  </tr>
</table>
<hr size="1"><address style="align: right;"><small>Generated on Wed Oct 11 16:34:18 2006 for PHAST by&nbsp;
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border="0"></a> 1.3.9.1 </small></address>
</body>
</html>
