<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<html><head><meta http-equiv="Content-Type" content="text/html;charset=iso-8859-1">
<title>msa.c File Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css">
</head><body>
<!-- Generated by Doxygen 1.2.18 -->
<center>
<a class="qindex" href="index.html">Main Page</a> &nbsp; <a class="qindex" href="modules.html">Modules</a> &nbsp; <a class="qindex" href="annotated.html">Compound List</a> &nbsp; <a class="qindex" href="files.html">File List</a> &nbsp; <a class="qindex" href="functions.html">Compound Members</a> &nbsp; <a class="qindex" href="globals.html">File Members</a> &nbsp; </center>
<hr><h1>msa.c File Reference</h1>Multiple sequence alignments. 
<a href="#_details">More...</a>
<p>
<table border=0 cellpadding=0 cellspacing=0>
<tr><td></td></tr>
<tr><td colspan=2><br><h2>Functions</h2></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a3">msa_new</a> (char **seqs, char **names, int nseqs, int length, char *alphabet)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Creates a new <a class="el" href="structMSA.html">MSA</a> object.</em> <a href="#a3"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a4">msa_new_from_file</a> (FILE *F, <a class="el" href="msa_8h.html#a66">msa_format_type</a> format, char *alphabet)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Creates a new alignment from the contents of the specified file, which is assumed to use the specified format.</em> <a href="#a4"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a5">msa_create_copy</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int suff_stats_only)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>create a copy of an <a class="el" href="structMSA.html">MSA</a>.</em> <a href="#a5"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a7">msa_print</a> (FILE *F, <a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="msa_8h.html#a66">msa_format_type</a> format, int pretty_print)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Prints <a class="el" href="structMSA.html">MSA</a> to file, using specified format.</em> <a href="#a7"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a9">msa_strip_gaps</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int gap_strip_mode)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>If gap_strip_mode is STRIP_ALL_GAPS or STRIP_ANY_GAPS, removes all columns with ALL or ANY gaps, respectively.</em> <a href="#a9"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a19" doxytag="msa.c::msa_get_seq_idx"></a>
int&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a19">msa_get_seq_idx</a> (<a class="el" href="structMSA.html">MSA</a> *msa, char *name)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Return sequence index of given sequence name or -1 if not found.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a28">msa_reverse_compl_feats</a> (<a class="el" href="structMSA.html">MSA</a> *msa, <a class="el" href="structGFF__Set.html">GFF_Set</a> *feats, int *aux_data)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Reverse complement segments of an <a class="el" href="structMSA.html">MSA</a> corresponding to groups of features on the reverse strand.</em> <a href="#a28"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a44" doxytag="msa.c::msa_format_for_suffix"></a>
<a class="el" href="msa_8h.html#a66">msa_format_type</a>&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a44">msa_format_for_suffix</a> (char *fname)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Return format type indicated by filename suffix.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a45" doxytag="msa.c::msa_suffix_for_format"></a>
char *&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a45">msa_suffix_for_format</a> (<a class="el" href="msa_8h.html#a66">msa_format_type</a> t)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Return appropriate filename suffix for format type.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top><a name="a48" doxytag="msa.c::msa_missing_col"></a>
int&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a48">msa_missing_col</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int ref, int pos)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Returns TRUE if alignment has missing data in all seqs but the reference seq at specified column; otherwise returns FALSE.</em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>List *&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a49">msa_seq_indices</a> (<a class="el" href="structMSA.html">MSA</a> *msa, List *seqnames)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Given a list of sequence names and/or 1-based indices, return a list of corresponding 0-based indices.</em> <a href="#a49"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a50">msa_mask_macro_indels</a> (<a class="el" href="structMSA.html">MSA</a> *msa, int k)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Mask out all alignment gaps of length greater than k by changing gap characters to missing data characters.</em> <a href="#a50"></a><em></em></font><br><br></td></tr>
<tr><td nowrap align=right valign=top>void&nbsp;</td><td valign=bottom><a class="el" href="msa_8c.html#a51">msa_set_informative</a> (<a class="el" href="structMSA.html">MSA</a> *msa, List *not_informative)</td></tr>
<tr><td>&nbsp;</td><td><font size=-1><em>Set up array indicating which sequences are to be considered "informative", e.g., for phylogenetic analysis.</em> <a href="#a51"></a><em></em></font><br><br></td></tr>
</table>
<hr><a name="_details"></a><h2>Detailed Description</h2>
Multiple sequence alignments.
<p>

<p>
<hr><h2>Function Documentation</h2>
<a name="a5" doxytag="msa.c::msa_create_copy"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structMSA.html">MSA</a>* msa_create_copy </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>suff_stats_only</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
create a copy of an <a class="el" href="structMSA.html">MSA</a>.
<p>
If suff_stats_only == 1, then sequences aren't copied     </td>
  </tr>
</table>
<a name="a50" doxytag="msa.c::msa_mask_macro_indels"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void msa_mask_macro_indels </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>k</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Mask out all alignment gaps of length greater than k by changing gap characters to missing data characters.
<p>
Useful when modeling micro-indels. Warning: if <a class="el" href="structMSA.html">MSA</a> is stored only in terms of sufficient statistics, an explicit alignment will be created     </td>
  </tr>
</table>
<a name="a3" doxytag="msa.c::msa_new"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structMSA.html">MSA</a>* msa_new </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">char **&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>seqs</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char **&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>names</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>nseqs</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>length</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>alphabet</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Creates a new <a class="el" href="structMSA.html">MSA</a> object.
<p>
Two-dimensional character arrays must be passed in for sequences and names (no new memory is allocated for them). The alphabet, however, will be copied into newly allocated memory. If the "alphabet" argument is null, the default alphabet will be used.     </td>
  </tr>
</table>
<a name="a4" doxytag="msa.c::msa_new_from_file"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> <a class="el" href="structMSA.html">MSA</a>* msa_new_from_file </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">FILE *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>F</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap><a class="el" href="msa_8h.html#a66">msa_format_type</a>&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>format</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>char *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>alphabet</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Creates a new alignment from the contents of the specified file, which is assumed to use the specified format.
<p>
If "alphabet" is NULL, default alphabet for DNA will be used. This routine will abort if the sequence contains a character not in the alphabet.     </td>
  </tr>
</table>
<a name="a7" doxytag="msa.c::msa_print"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void msa_print </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top">FILE *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>F</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap><a class="el" href="msa_8h.html#a66">msa_format_type</a>&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>format</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>pretty_print</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Prints <a class="el" href="structMSA.html">MSA</a> to file, using specified format.
<p>
The "pretty_print" option causes periods ('.') to be printed in place of characters that are identical to corresponding characters in the first sequence.     </td>
  </tr>
</table>
<a name="a28" doxytag="msa.c::msa_reverse_compl_feats"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void msa_reverse_compl_feats </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap><a class="el" href="structGFF__Set.html">GFF_Set</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>feats</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>aux_data</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Reverse complement segments of an <a class="el" href="structMSA.html">MSA</a> corresponding to groups of features on the reverse strand.
<p>
Adjusts the coordinates in the <a class="el" href="structGFF__Set.html">GFF_Set</a> accordingly. This function can be used to ensure that sites in strand-specific categories (e.g., 1st codon position, intron) are oriented consistently. It can be useful in phylogenetic analysis and in training the transition probabilities of a phylo-HMM. Features are assumed already to have been grouped as desired, and groups are assumed to be nonoverlapping (see gff_group and gff_remove_overlaps). Strandedness is tested using gff_reverse_strand_only. The <a class="el" href="structGFF__Set.html">GFF_Set</a> is assumed to use the coordinate frame of the alignment. <dl compact><dt><b>Parameters: </b></dt><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>msa</em>&nbsp;</td><td>
Alignment object. If NULL, only the <a class="el" href="structGFF__Set.html">GFF_Set</a> (and optionally aux_data) will be altered </td></tr>
<tr><td valign=top><em>feats</em>&nbsp;</td><td>
Set of features </td></tr>
<tr><td valign=top><em>aux_data</em>&nbsp;</td><td>
Auxiliary array of site-specific integers (e.g., gap patterns) to be kept in sync with the alignment and/or <a class="el" href="structGFF__Set.html">GFF_Set</a> </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a49" doxytag="msa.c::msa_seq_indices"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> List* msa_seq_indices </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>List *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>seqnames</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Given a list of sequence names and/or 1-based indices, return a list of corresponding 0-based indices.
<p>
Abort if a name has no match. Useful in converting command-line arguments     </td>
  </tr>
</table>
<a name="a51" doxytag="msa.c::msa_set_informative"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void msa_set_informative </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>List *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>not_informative</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
Set up array indicating which sequences are to be considered "informative", e.g., for phylogenetic analysis.
<p>
<dl compact><dt><b>Parameters: </b></dt><dd>
<table border=0 cellspacing=2 cellpadding=0>
<tr><td valign=top><em>msa</em>&nbsp;</td><td>
Alignment </td></tr>
<tr><td valign=top><em>not_informative</em>&nbsp;</td><td>
List of names of sequences *not* to be considered informative </td></tr>
</table>
</dl>    </td>
  </tr>
</table>
<a name="a9" doxytag="msa.c::msa_strip_gaps"></a><p>
<table width="100%" cellpadding="2" cellspacing="0" border="0">
  <tr>
    <td class="md">
      <table cellpadding="0" cellspacing="0" border="0">
        <tr>
          <td class="md" nowrap valign="top"> void msa_strip_gaps </td>
          <td class="md" valign="top">(&nbsp;</td>
          <td class="md" nowrap valign="top"><a class="el" href="structMSA.html">MSA</a> *&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>msa</em>, </td>
        </tr>
        <tr>
          <td></td>
          <td></td>
          <td class="md" nowrap>int&nbsp;</td>
          <td class="mdname" nowrap>&nbsp; <em>gap_strip_mode</em></td>
        </tr>
        <tr>
          <td></td>
          <td class="md">)&nbsp;</td>
          <td class="md" colspan="2"></td>
        </tr>

      </table>
    </td>
  </tr>
</table>
<table cellspacing=5 cellpadding=0 border=0>
  <tr>
    <td>
      &nbsp;
    </td>
    <td>

<p>
If gap_strip_mode is STRIP_ALL_GAPS or STRIP_ANY_GAPS, removes all columns with ALL or ANY gaps, respectively.
<p>
Otherwise, assumes a projection* is desired onto the sequence whose index is gap_strip_mode (indexing starts with 1). Changes are made to original alignment. Gaps are expected to be represented by GAP_CHAR. If msa-&gt;categories is non-NULL, will be adjusted accordingly.     </td>
  </tr>
</table>
<hr><address style="align: right;"><small>Generated on Fri Jun 24 14:59:01 2005 for PHAST by
<a href="http://www.doxygen.org/index.html">
<img src="doxygen.png" alt="doxygen" align="middle" border=0 
width=110 height=53></a>1.2.18 </small></address>
</body>
</html>
